contracts/NFTOpt.sol

```diff
@@ -49,13 +68,29 @@ contract NFTOpt {
     external
     payable
     {
+        require(options[_optionId].buyer != address(0), "Option with the specified id does not exist");
+        require(options[_optionId].seller == address(0), "Option is already fulfilled by a seller");
+        require(options[_optionId].state == OptionState.REQUEST, "Option is not in the request state");
+        require(options[_optionId].buyer != msg.sender, "Seller is the same as buyer");
+        require(address(msg.sender).balance >= options[_optionId].strikePrice, "Seller does not have enough balance");
+        require(getBalance() <= options[_optionId].premium, "Not enough funds to pay the premium to the seller");
+
+        options[_optionId].seller = msg.sender;
+        options[_optionId].startDate = block.timestamp;
+        options[_optionId].state = OptionState.OPEN;
+
+        emit Filled(msg.sender, _optionId);

```

Comment:
this should be LAST; switch with lines 84-85 (because if the `call` invocation fails, you would still emit the event)